nixops/backends/ec2.py

```diff
@@ -57,7 +57,7 @@ def __init__(self, xml, config):
         self.associate_public_ip_address = config["ec2"]["associatePublicIpAddress"]
         self.use_private_ip_address = config["ec2"]["usePrivateIpAddress"]
         self.security_group_ids = config["ec2"]["securityGroupIds"]
-        self.block_device_mapping = {_xvd_to_sd(k): v for k, v in config["ec2"]["blockDeviceMapping"].iteritems()}
+        self.block_device_mapping = config["ec2"]["blockDeviceMapping"]
```

Comment:
while it generally looks fine, have you checked whether there are backwards compatibility problems?

I think the situation is like this. Early NixOps version used `/dev/sdaX` to nominate devices, but Amazon later renamed those to `/dev/xvdX`, so NixOps did a hack to runtime replace `/dev/sdX -> /dev/xvdX` for `nixops.state` config. So in theory, there may be NixOps installations with `/dev/sdX` entries in `blockDeviceMapping` config. 

This and further usages of `self.block_device_mapping` may work incorrectly in this scenario. Though I haven't checked this thoroughly, I'll leave it to you :)

Reply:
I do

```sh
 ~/projects/nixops   fix-aws-5gen-disk-attach  gco origin/master
Note: checking out 'origin/master'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 3b2751e Merge pull request #933 from amemni/gce-labels-disks-snaps
 ~/projects/nixops  ➦ 3b2751e 

~/projects/vd-rails-deploy   master ✚  nix-shell
....build

[nix-shell:~/projects/vd-rails-deploy]$ make nixops_create
nixops create '<base.nix>'
created deployment ‘daeb8e51-6dab-11e8-8272-02425e3939e4’
daeb8e51-6dab-11e8-8272-02425e3939e4


[nix-shell:~/projects/vd-rails-deploy]$ nixops deploy
......
backend.............> attaching volume ‘vol-09b9b99d682509147’ as ‘/dev/xvdf’... [attaching] [attached]
building all machine configurations...
.............

backend.............> A dependency job for local-fs.target failed. See 'journalctl -xe' for details.
backend.............> error: Traceback (most recent call last):
  File "/nix/store/6pzj7d648mksqdmk1npd6ljrk59xbz2p-python2.7-nixops-1.6.1pre0_abcdef/lib/python2.7/site-packages/nixops/deployment.py", line 731, in worker
    raise Exception("unable to activate new configuration (exit code {})".format(res))
Exception: unable to activate new configuration (exit code 4)

error: activation of 1 of 1 machines failed (namely on ‘backend’)

[nix-shell:~/projects/vd-rails-deploy]$ nixops ssh backend

[root@backend:~]# systemctl status local-fs.target
● local-fs.target - Local File Systems
   Loaded: loaded (/nix/store/6dkz6azki0bag13m6vccy7bi41n9dak3-systemd-237/example/systemd/system/local-fs.target; enabled; vendor preset: enabled)
  Drop-In: /nix/store/r8m7q7j1sd854cqh50yjq6xymmldbrlq-system-units/local-fs.target.d
           └─overrides.conf
   Active: inactive (dead) since Mon 2018-06-11 19:19:37 UTC; 4min 53s ago
     Docs: man:systemd.special(7)

Jun 11 19:19:37 ip-172-31-73-78.ec2.internal systemd[1]: Stopped target Local File Systems.
Jun 11 19:21:14 backend systemd[1]: Dependency failed for Local File Systems.
Jun 11 19:21:14 backend systemd[1]: local-fs.target: Job local-fs.target/start failed with result 'dependency'.
Jun 11 19:21:14 backend systemd[1]: local-fs.target: Triggering OnFailure= dependencies.
Jun 11 19:21:14 backend systemd[1]: local-fs.target: Failed to enqueue OnFailure= job: No such file or directory

[root@backend:~]# systemctl status dev-sdf.device
● dev-sdf.device
   Loaded: loaded
   Active: inactive (dead)

Jun 11 19:21:14 backend systemd[1]: dev-sdf.device: Job dev-sdf.device/start timed out.
Jun 11 19:21:14 backend systemd[1]: Timed out waiting for device dev-sdf.device.
Jun 11 19:21:14 backend systemd[1]: dev-sdf.device: Job dev-sdf.device/start failed with result 'timeout'.
```

change to 

```
      fileSystems."/data" = {
        autoFormat = true;
        fsType = "ext4";
        device = "/dev/xvdf";
        ec2.disk = resources.ebsVolumes.foo-disk;
      };
```


[nix-shell:~/projects/vd-rails-deploy]$ nixops deploy
building all machine configurations...
these derivations will be built:
....

backend.............> activation finished successfully
foo> deployment finished successfully

```

```
~/projects/nixops  ➦ 3b2751e  git checkout fix-aws-5gen-disk-attach
Previous HEAD position was 3b2751e Merge pull request #933 from amemni/gce-labels-disks-snaps
Switched to branch 'fix-aws-5gen-disk-attach'
Your branch is up to date with 'origin/fix-aws-5gen-disk-attach'.

```


```
~/projects/vd-rails-deploy   master ●✚  nix-shell

[nix-shell:~/projects/vd-rails-deploy]$ nixops deploy
backend.............> attaching volume ‘vol-09b9b99d682509147’ as ‘/dev/xvdf’...
building all machine configurations...
backend.............> copying closure...
foo> closures copied successfully
backend.............> updating GRUB 2 menu...
backend.............> activating the configuration...
backend.............> setting up /etc...
backend.............> setting up tmpfiles
backend.............> activation finished successfully
backend.............> detaching device ‘/dev/sdf’...
backend.............> umount: /dev/sdf: no mount point specified.
foo> deployment finished successfully

[nix-shell:~/projects/vd-rails-deploy]$ nixops ssh backend
Last login: Mon Jun 11 19:22:45 2018 from 185.29.253.219

[root@backend:~]# lsblk
NAME    MAJ:MIN RM SIZE RO TYPE MOUNTPOINT
xvda    202:0    0   5G  0 disk
└─xvda1 202:1    0   5G  0 part /
xvdf    202:80   0   5G  0 disk /data


```


P.S.

I wrote about this here https://github.com/NixOS/nixops/pull/965#issuecomment-394267539

Reply:
I see, yes, no problem here. I was talking about the following configuration:
```
let

  region = "us-west-1";
  accessKeyId = "default";
  ec2 =
    { resources, ... }:
    { deployment.targetEnv = "ec2";
      deployment.ec2.accessKeyId = accessKeyId;
      deployment.ec2.region = region;
      deployment.ec2.instanceType = "m3.medium";
      deployment.ec2.keyPair = resources.ec2KeyPairs.my-key-pair;

      deployment.ec2.blockDeviceMapping."/dev/sdc" = {
          disk = "ephemeral0";
      };
    };

in
{ proxy    = ec2;

  # Provision an EC2 key pair.
  resources.ec2KeyPairs.my-key-pair =
    { inherit region accessKeyId; };
}
```
And turns out that due to how `boto` works, it is fine:
```
[root@proxy:~]# df -h
Filesystem                Size  Used Avail Use% Mounted on
devtmpfs                  189M     0  189M   0% /dev
tmpfs                     1.9G     0  1.9G   0% /dev/shm
tmpfs                     941M  3.9M  937M   1% /run
tmpfs                     1.9G  300K  1.9G   1% /run/wrappers
/dev/disk/by-label/nixos  3.0G  1.6G  1.3G  56% /
/dev/xvdc                 3.9G  8.1M  3.7G   1% /disk0
tmpfs                     1.9G     0  1.9G   0% /sys/fs/cgroup
tmpfs                     377M     0  377M   0% /run/user/0
```
even though `/dev/sdX` don't work anymore for EC2

Reply:
Yep, still good that we have tested it)